posts/xxxx-xx-xx-how-reuters-replaced-websockets-with-amazon-cognito-and-sqs.md

```diff
@@ -0,0 +1,150 @@
+
+The only solution that works at an arbitrarily high scale is to grant message read and delete access to all queues based on the name prefix of the queue.  The queue names are generated using a random 256 bit value generated by the Linux kernel using environmental entropy [/dev/urandom](https://linux.die.net/man/4/urandom), which is suitable for generating unguessable and secure Queue URLs.  Another 256 random bits are generated for an AES key that is used to encrypt and decrypt all messages.  This key is returned to the caller when creating a session and stored in the DynamoDB session table.  A random counter initialization value is used for the AES stream cipher and returned along with the encrypted message.  The *decrypt-message* script in the project provides an example Python implementation of decryption and a Javascript version is provided in the project's [documentation](https://github.com/ReutersMedia/sqs-browser-events).
+
+Any set of Cognito credentials from the pool can be used to read and delete messages from any queue.  The Cognito [GetId method](http://docs.aws.amazon.com/cognitoidentity/latest/APIReference/API_GetId.html) is also public, so anyone with knowledge of the Cognito endpoint can generate ID's.  However, to access a Queue, one would need to know the random Queue name and URL. If they know the Queue URL, they can delete messages, but can't decrypt them without the private key.  Whether this provides sufficient security depends on the nature of the data being handled.  We hope however that AWS will at some point provide the tools necessary to secure a queue to a specific Cognito identity.
+
+Access to the user message API Gateway endpoints is secured via IAM, meaning users can't directly call them.  In our case, access is mediated through a separate service that checks the user's credentials.
+
+## Handling High Message Volumes
+
+Dispatching messages to individual users is relatively easy to scale if they are injected through the API Gateway.  Messages that have to be broadcast to all users, or to accounts with a large number of users, can become more difficult.  Each message requires an AWS API call to SQS to insert into the queue, and a call to DynamoDB to add it to the user's message history.  Because these operations are mostly io wait, we use Lambda's with low memory allocations of 128MB or 256MB, making operation as economical as possible.  Multi-threading the operations, the [dispatcher Lambda](https://github.com/ReutersMedia/sqs-browser-events/blob/master/dispatcher.py) can only get off a few thousand API calls before timing out.  If you need to broadcast a breaking news event to 100,000 users, a single Lambda invocation will not suffice.
```

Comment:
@kaellis  "are mostly io wait", what exactly are you saying here? mind making this a little more clear?